src/nbl/ui/CWindowXCB.cpp

```diff
@@ -0,0 +1,296 @@
+    wnd->setWindowVisible_impl(visible);
+    return true;
+}
+
+bool CWindowManagerXCB::setWindowMaximized_impl(IWindow* window, bool maximized) {
+    auto wnd = static_cast<IWindowXCB*>(window);
+    wnd->setWindowMaximized_impl(maximized);
+    return true;
+}
+
+CWindowXCB::CWindowXCB(core::smart_refctd_ptr<CWindowManagerXCB>&& winManager, SCreationParams&& params):
+    IWindowXCB(std::move(params)),
+    m_windowManager(winManager),
+    m_connection(core::make_smart_refctd_ptr<XCBConnection>(core::smart_refctd_ptr<CWindowManagerXCB>(m_windowManager))),
+    m_cursorControl(core::make_smart_refctd_ptr<CCursorControlXCB>(core::smart_refctd_ptr<XCBConnection>(m_connection))),
+    m_clipboardManager(core::make_smart_refctd_ptr<CClipboardManagerXCB>(core::smart_refctd_ptr<XCBConnection>(m_connection))),
+    m_dispatcher(*this) {
+    
+    auto& xcb = m_windowManager->getXcbFunctionTable();
+    auto& xcbIccm = m_windowManager->getXcbIcccmFunctionTable();
+
+    m_xcbWindow = xcb.pxcb_generate_id(m_connection->getRawConnection());
+
+    const auto* primaryScreen = m_connection->primaryScreen();
+
+    uint32_t eventMask = XCB_CW_BACK_PIXEL | XCB_CW_EVENT_MASK;
+    uint32_t valueList[] = {
+        primaryScreen->black_pixel,
+        XCB_EVENT_MASK_STRUCTURE_NOTIFY | XCB_EVENT_MASK_KEY_PRESS | XCB_EVENT_MASK_KEY_RELEASE |
+            XCB_EVENT_MASK_FOCUS_CHANGE | XCB_EVENT_MASK_PROPERTY_CHANGE
+    };
+
+    xcb_void_cookie_t xcbCheckResult = xcb.pxcb_create_window(
+        m_connection->getRawConnection(), XCB_COPY_FROM_PARENT, m_xcbWindow, primaryScreen->root,
+        static_cast<int16_t>(m_x),
+        static_cast<int16_t>(m_y),
+        static_cast<int16_t>(m_width),
+        static_cast<int16_t>(m_height), 4,
+        XCB_WINDOW_CLASS_INPUT_OUTPUT, primaryScreen->root_visual, eventMask,
+        valueList);
+
+    setWindowSize_impl(m_width, m_height);
+    
+    auto WM_DELETE_WINDOW = m_connection->resolveAtom(m_WM_DELETE_WINDOW);
+    auto NET_WM_PING = m_connection->resolveAtom(m_NET_WM_PING);
+    auto WM_PROTOCOLS = m_connection->resolveAtom(m_WM_PROTOCOLS);
+    
+    const std::array atoms {WM_DELETE_WINDOW, NET_WM_PING};
+    xcb.pxcb_change_property(
+            m_connection->getRawConnection(), 
+            XCB_PROP_MODE_REPLACE, 
+            m_xcbWindow, 
+            WM_PROTOCOLS, XCB_ATOM_ATOM, 32, atoms.size(), atoms.data());
+
+
+    auto motifHints = fetchMotifMWHints(getFlags().value);
+    m_connection->setMotifWmHints(m_xcbWindow, motifHints);
+    
+    if(isAlwaysOnTop()) {
+        XCBConnection::XCBAtomToken<core::StringLiteral("NET_WM_STATE_ABOVE")> NET_WM_STATE_ABOVE;
+        m_connection->setNetMWState(
+            primaryScreen->root,
+            m_xcbWindow, false, m_connection->resolveAtom(NET_WM_STATE_ABOVE));
+    }
+
+    xcb.pxcb_map_window(m_connection->getRawConnection(), m_xcbWindow);
+    xcb.pxcb_flush(m_connection->getRawConnection());
+    m_dispatcher.start();
```

Comment:
there's a fun trick, `IThreadHandler` has a special constructor with a tag parameter which tells it to start the thread on construction